File: api/v1/account/account.go
package account

import (
	"bytes"
	"io/ioutil"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/meshify-app/meshify/auth"
	core "github.com/meshify-app/meshify/core"
	model "github.com/meshify-app/meshify/model"
	log "github.com/sirupsen/logrus"
	"golang.org/x/oauth2"
)

// ApplyRoutes applies router to gin Router
func ApplyRoutes(r *gin.RouterGroup) {
	g := r.Group("/accounts")
	{

		g.POST("/", createAccount)
		g.POST("/:id/activate", activateAccount)
		g.PATCH("/:id/activate", activateAccount)
		g.GET("/:id", readAllAccounts)
		g.PATCH("/:id", updateAccount)
		g.DELETE("/:id", deleteAccount)
	}
}

func activateAccount(c *gin.Context) {
	id := c.Param("id")

	v, err := core.ActivateAccount(id)
	if err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to create account")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, v)
}

func createAccount(c *gin.Context) {
	var account model.Account

	if err := c.ShouldBindJSON(&account); err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to bind")
		c.AbortWithStatus(http.StatusUnprocessableEntity)
		return
	}

	// get creation user from token and add to client infos
	oauth2Token := c.MustGet("oauth2Token").(*oauth2.Token)
	oauth2Client := c.MustGet("oauth2Client").(auth.Auth)
	user, err := oauth2Client.UserInfo(oauth2Token)
	if err != nil {
		log.WithFields(log.Fields{
			"oauth2Token": oauth2Token,
			"err":         err,
		}).Error("failed to get user with oauth token")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	account.From = user.Email

	v, err := core.CreateAccount(&account)
	if err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to create account")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	log.Infof("emailUser account = %v %v %v", account.Email, account.Id, account.MeshId)

	err = core.EmailUser(account.Email, account.Id, account.MeshId)
	if err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to send email to client")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, v)
}

func readAllAccounts(c *gin.Context) {
	email := c.Param("id")

	accounts, err := core.ReadAllAccounts(email)
	if err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to read accounts")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, accounts)
}

func updateAccount(c *gin.Context) {
	var data model.Account
	id := c.Param("id")

	var bodyBytes []byte
	if c.Request.Body != nil {
		bodyBytes, _ = ioutil.ReadAll(c.Request.Body)
		log.Info("updateAccount - %s", string(bodyBytes))
	}

	// Restore the io.ReadCloser to its original state
	c.Request.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))

	log.Info("updateAccount - %s", string(bodyBytes))

	if err := c.ShouldBindJSON(&data); err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to bind")
		c.AbortWithStatus(http.StatusUnprocessableEntity)
		return
	}
	client, err := core.UpdateAccount(id, &data)
	if err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to update client")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, client)
}

func deleteAccount(c *gin.Context) {
	id := c.Param("id")

	err := core.DeleteAccount(id)
	if err != nil {
		log.WithFields(log.Fields{
			"err": err,
		}).Error("failed to remove client")
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, gin.H{})
}

File: ui/src/store/modules/account.js
import ApiService from "../../services/api.service";

const state = {
  error: null,
  accounts: [],
  users: [],
  members: [],
}

const getters = {
  error(state) {
    return state.error;
  },
  accounts(state) {
    return state.accounts;
  },
  users(state) {
    return state.users;
  },
  members(state) {
    return state.members;
  },
}

const actions = {
  error({ commit }, error){
    commit('error', error)
  },

  readAll({ commit, dispatch }, id){
    ApiService.get(`/accounts/${id}`)
      .then(resp => {
        commit('accounts', resp)
      })
      .catch(err => {
        commit('error', err)
      })
  },

  readUsers({ commit, dispatch }, id){
    ApiService.get(`/accounts/${id}`)
      .then(resp => {
        commit('users', resp)
      })
      .catch(err => {
        commit('error', err)
      })
  },

  readMembers({ commit, dispatch }, id){
    ApiService.get(`/accounts/${id}`)
      .then(resp => {
        commit('members', resp)
      })
      .catch(err => {
        commit('error', err)
      })
  },

  create({ commit }, account){
    ApiService.post(`/accounts`, account)
      .then(resp => {
        commit('create', resp)
      })
      .catch(err => {
        commit('error', err)
      })
  },

  update({ commit, dispatch }, account){
    ApiService.patch(`/accounts/${account.id}`,account)
      .then(resp => {
        commit('update', resp)
      })
      .catch(err => {
        commit('error', err)
      })
  },

  delete({ commit }, account){
    ApiService.delete(`/accounts/${account.id}`)
      .then(() => {
        commit('delete', account)
      })
      .catch(err => {
        commit('error', err)
      })
  },

  email({ commit }, account){
    ApiService.get(`/accounts/${account.id}/invite`)
      .then(() => {
      })
      .catch(err => {
        commit('error', err)
      })
  },


}

const mutations = {
  error(state, error) {
    state.error = error;
  },
  accounts(state, accounts){
    state.accounts = accounts
  },
  users(state, users) {
    state.users = users
  },
  members(state, members) {
    state.members = members
  },
  create(state, account){
    state.accounts.push(account)
  },
  create(state, user) {
    state.users.push(user)
  },
  update(state, account){
    let index = state.accounts.findIndex(x => x.id === account.id);
    if (index !== -1) {
      state.accounts.splice(index, 1);
      state.accounts.push(account);
    } else {
      state.error = "update account failed, not in list"
    }
  },
  delete(state, account){
    let index = state.accounts.findIndex(x => x.id === account.id);
    if (index !== -1) {
      state.accounts.splice(index, 1);
    } else {
      state.error = "delete account failed, not in list"
    }
  },
  update(state, user){
    let index = state.users.findIndex(x => x.id === user.id);
    if (index !== -1) {
      state.users.splice(index, 1);
      state.users.push(user);
    } else {
      state.error = "update account (user) failed, not in list"
    }
  },
  delete(state, user){
    let index = state.users.findIndex(x => x.id === user.id);
    if (index !== -1) {
      state.users.splice(index, 1);
    } else {
      state.error = "delete user failed, not in list"
    }
  },
  update(state, member){
    let index = state.users.findIndex(x => x.id === member.id);
    if (index !== -1) {
      state.members.splice(index, 1);
      state.members.push(member);
    } else {
      state.error = "update account (member) failed, not in list"
    }
  },
  delete(state, member){
    let index = state.users.findIndex(x => x.id === member.id);
    if (index !== -1) {
      state.members.splice(index, 1);
    } else {
      state.error = "delete user failed, not in list"
    }
  },
}

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations
}
